### ISSUE TITLE
No line break insertions in short CSS selectors

### ISSUE THREAD

**Issue by janosh** [prettier/prettier#5086]
**Prettier 1.14.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLAtgBwgJxgAgGcYBPAGzgBN8AzHCdfAcmPKoFpItoEZCmAOlCEZsefOiqoAhgEUArnBwla9RkwB0GgPRbt8mKjKFtkyjIVKSg4VEhRi+AArS8qMKkzTYhfAF4iUgpKDUI4MENoAAMhfHwvSnMoAHMkfAAmHDh0AG5Y-AAjaTAAa2T6eShKNIASYEx6TF8-AD54xsINGAALbLgNdGlUKAB1btR4AF88qDjzQkwyaRI08tRKGbi1ynZk6Uw0zOzN-G6ARgAaU-Sr7oBmW4AWfGB8uIB3dZ60s4AGX4ApCc4gkkql8L8NABWLKMI65N6FYplCpVWr1Dr+NoNCBNLq9SQaMioZLdGAAcSyCGm+Um+TqZgsimUGkw3R4kxeiPOtxupwep2er1mcVFNGgMHYhFQAC84D9YcD8HSRSq4gypHJmSQutIChQYJzhaL8NtdvsFcduZdrrcBd0hYitjh1pwIGR5OgoD9tOwzkq1cr6cBGVqrANhgAVPUGo3W3l2p5ckUm97jeBSrxgeX4KAQd44fZK52u+jvQ4B2lCGK2OAADzEBEocBo0nkZAILjcHi8PhAFxAuMiDmQoFcZa7CEIyBA0gAbhB1v2QAVC6U4DAAMpZ4bJZAwHCKAfDMJ4JyF5KDZCt4xwAcAK0IdYAQmuShvN9JJAAZYZwa-SLeA5eDgp4zkUBTkNAy4NMMMAjF83TIAAHL8wH0GEIyFpgM4NHAp5zv+A5ZAAjvIqBZOe0iXtIAFASAYToKgdFhAO0opBQCgQPALF3iAMB6ghlA9MgNz8YWRi7gAwgwV4oHmUBESA8hhNGBTTkgN5hJMkxAA)
```sh
--parser babylon
--no-semi
```

I suggest reconfiguring the changes made in #1962 so as to only insert line breaks in CSS selectors, when they exceed a certain length. Otherwise, you end up with an unnecessarily and unreadably long file in cases such as the one below.

**Input:**
```jsx
import styled from 'styled-components'

import mediaQuery from '../../utils/mediaQuery'

const Participants = styled.section`
  padding: 2rem;
  background: ${props => props.theme.mainWhite};
  display: grid;
  grid-gap: 2rem;
  h1, h2, h3, h4 {
    width: 100%;
    padding: 0.5rem 2rem;
    background: ${props => props.theme.lightGreen};
  }
  ${mediaQuery.phone} {
    h1, h2, h3, h4 {
      font-size: 1rem;
    }
  }
  ${mediaQuery.tablet} {
    grid-gap: 1rem;
    h1, h2, h3, h4 {
      grid-column: 1/-1;
    }
  }
  ${mediaQuery.minTablet} {
    h1, h2, h3, h4 {
      white-space: nowrap;
      grid-row: 2;
    }
  }
`

export default Participants
```

**Output:**
```jsx
import styled from "styled-components"

import mediaQuery from "../../utils/mediaQuery"

const Participants = styled.section`
  padding: 2rem;
  background: ${props => props.theme.mainWhite};
  display: grid;
  grid-gap: 2rem;
  h1,
  h2,
  h3,
  h4 {
    width: 100%;
    padding: 0.5rem 2rem;
    background: ${props => props.theme.lightGreen};
  }
  ${mediaQuery.phone} {
    h1,
    h2,
    h3,
    h4 {
      font-size: 1rem;
    }
  }
  ${mediaQuery.tablet} {
    grid-gap: 1rem;
    h1,
    h2,
    h3,
    h4 {
      grid-column: 1/-1;
    }
  }
  ${mediaQuery.minTablet} {
    h1,
    h2,
    h3,
    h4 {
      white-space: nowrap;
      grid-row: 2;
    }
  }
`

export default Participants

```

**Expected behavior:**

No insertion of line breaks in CSS selectors of length less than,  say, 80.

**Comment by j-f1** [comment]
Perhaps we shouldn’t break if all of the sectors are bare element selectors.

**Comment by lydell** [comment]
@j-f1 Sounds reasonably. Thinking back of the previous issues, I think that’s the primary use case.

**Comment by janosh** [comment]
Just out of curiosity, was this issue addressed by any PR?

**Comment by lydell** [comment]
@janosh No PR for this so far!

**Comment by j-f1** [comment]
Let’s go with the heuristic of not automatically breaking selectors that consist of just element selectors.

**Comment by SYNTAG** [comment]
> Let’s go with the heuristic of not automatically breaking selectors that consist of just element selectors.

This would undeniably be better default behavior than the current implementation. And since this is a relevant topic, I still think it's worth considering offering a config for this, whether it be toggling multiline for style selectors or setting some kind of `max-len` as referenced here [#5937](https://github.com/prettier/prettier/issues/5937)

It can sometimes be more preferable to read/write:
```css
.heading, h1, h2 {
   margin: 0;
}
```
as opposed to:
```css
.heading
h1,
h2 {
  margin: 0;
}
```

Prettier has handled for this in JavaScript, so what I am proposing is something similar for CSS too. For example, it's generally more preferable to read/write the following:
```javascript
const { Header, Footer, Nav } = require('./layout');
```

as opposed to:
```javascript
const {
  Header,
  Footer,
  Nav,
} = require('./layout');
```

**Comment by lydell** [comment]
I think we should keep things as they are. Keep it simple! As of this writing, there are no :+1:s on this issue.

**Comment by evilebottnawi** [comment]
I think we can close issue, printing in one line unreadable

**Comment by SYNTAG** [comment]
> I think we can close issue, printing in one line unreadable

To say the one-line code examples given above are unreadable is a bit of an over exaggeration.

I think @j-f1 idea would be a great solution to this.

**Comment by lydell** [comment]
Feel free to provide more examples! :+1:

**Comment by SYNTAG** [comment]
Here are some more real examples of scenarios where it's nicer to have one line:

```css
*, *:before, *:after {
  box-sizing: border-box;
}
```
```css
ul, ol {
    margin: 0;
    padding: 0;
    list-style: none;
}
```

```css
h1, h2, h3, h4, h5, h6 {
  margin: 0;
}

There are a lot of cases like this where it's simpler to keep things compact. Putting each selector on its own line is nice, but only really necessary when you start selecting more things.

**Comment by janosh** [comment]
@lydell I'm a little surprised that this issue is even debated. I was under the impression that the point of prettier is not to be simple in and of itself but to produce ideally formatted code. I can't imagine anyone would consider this to be ideally formatted,
```css
h1,
h2,
h3,
h4,
h5, 
h6 {
  margin: 0;
}
```
especially when compared to this
```css
h1, h2, h3, h4, h5, h6 {
  margin: 0;
}
```
Breaking code onto multiple lines only once it exceeds a certain length is such a common practice. I don't see why it would be more difficult or less useful with CSS selectors than anywhere else.

**Comment by lydell** [comment]
I think Prettier already has too many heuristics that break out of the simple rules. For example, #5939. What usually happens when we add stuff like this is that people open a new issue wondering why their code is/isn't broken across several lines in some cases.

I see what you mean with the `h1, h2, h3, h4, h5, h6` and `ul, ol` examples. I've experienced those too in basically all projects I've worked on. But I only write those selectors one or two times in the entire project, so it didn't bother me much.

But maybe this is fine to add. I guess I'm just a little more hesitant to changes since the recent ternary change backlash: #5814 (for example). Basically every time we change something new people come and want it changed back. And this issue only has 2 :+1:s so far.

**Comment by janosh** [comment]
@lydell I certainly understand the reluctance to introduce additional heuristics. Just to be clear though, I wasn't advocating for one. I'm suggesting to only break CSS selectors onto multiple lines if they exceed a certain length _in general_. Is there a reason why this isn't the current behavior?

**Comment by lydell** [comment]
Ah, I see. It was decided early on to print every selector on its own line: #2057

**Comment by janosh** [comment]
Thanks for digging that up! Interestingly, that issue also didn't receive any upvotes. Seems like there is low interest in CSS selectors in general. FWIW, I believe making that change was a misstep.

**Comment by lydell** [comment]
Regarding upvotes – that issue is from when Prettier was not even half a year old, and the CSS support was even younger than that. Things were moving quickly, and since then the bar for changes has grown increasingly higher.

**Comment by sublimeye** [comment]
So is it possible to reconfigure this behavior?

**Comment by lydell** [comment]
@sublimeye There's no option for this currently.

**Comment by janosh** [comment]
I think @sublimeye may have been referring to reverting #2057, i.e. only breaking CSS selectors onto multiple lines if they exceed a certain line length.

**Comment by lydell** [comment]
There are no plans to change this currently, either.

**Comment by SYNTAG** [comment]
@lydell It's a shame really. I understand that you're looking for an overwhelming consensus before taking any action since you probably have more important things to focus on.

But with Preitter being an already opinionated formatter, I was at least hoping that even though this thread lacks popularity, that the change request shares similar values that were already made on other languages for virtually the same scenario in Javascript. I don't understand why SCSS is being treated differently.

I'd least hope that we can look at this change request objectively and see that implementing it would not only make sense for Preitter, but add consistency across different languages.

**Comment by evilebottnawi** [comment]
@SYNTAG instead says shame you can help, we help you on a voluntary basis, and it’s very unpleasant to hear it

**Comment by SYNTAG** [comment]
@evilebottnawi I didn't mean to come off rude or disrespectful. I appreciate all that you do, and I understand how it is to work on a volunteer basis. I would certainly help if I had the capabilities, but I'm still fairly to new Javascript.

I'm just very passionate about using Prettier for my projects and was hoping to guide this change request in a direction where it might be considered for a future update, despite its lack of popularity.

**Comment by fisker** [comment]
Code Guide suggest:

> When grouping selectors, keep individual selectors to a single line.

https://codeguide.co/#css

`stylelint-config-standard` has

`"selector-list-comma-newline-after": "always",`

https://github.com/stylelint/stylelint-config-standard/blob/5d1531025526324abdfc6d1fbf40b5fb9f2d7c2f/index.js#L95

`normalize.css` follows the same style

https://github.com/necolas/normalize.css/blob/fc091cce1534909334c1911709a39c22d406977b/normalize.css#L125-L126

`bootstrap`

https://github.com/twbs/bootstrap/blob/66e9ab3e4dd0a5fe501e8499e45bbd7488b80fd0/scss/_reboot.scss#L14-L16

https://github.com/search?q=h3+repo%3Atwbs%2Fbootstrap+extension%3Ascss&type=Code

**Comment by aravindvnair99** [comment]
I feel it makes better sense to split to multiple lines when it goes beyond a certain number of characters as pointed out by @SYNTAG at https://github.com/prettier/prettier/issues/5086#issuecomment-469507962, @janosh at https://github.com/prettier/prettier/issues/5086#issue-359655142, @j-f1 at https://github.com/prettier/prettier/issues/5086#issuecomment-420960646

Also, as per the official [Prettier - Options](https://prettier.io/docs/en/options.html#print-width) page, it says

> For readability we recommend against using more than 80 characters:
> 
> In code styleguides, maximum line length rules are often set to 100 or 120. However, when humans write code, they don't strive to reach the maximum number of columns on every line. Developers often use whitespace to break up long lines for readability. In practice, the average line length often ends up well below the maximum.
> 
> Prettier, on the other hand, strives to fit the most code into every line. With the print width set to 120, prettier may produce overly compact, or otherwise undesirable code.

So why not apply the same logic to CSS as well? Put as many as 80 characters and only split when the code is untidy instead of splitting for each selector which makes it tough to scroll or decreases readability in the case of long CSS files. I feel an option to choose between the two styles would be nice to have.

**Comment by Jason-2020** [comment]
Is there any workaround?

**Comment by nickbclifford** [comment]
Hello from 2021! Have there been any updates or workarounds for this since the last post?

**Comment by janosh** [comment]
@nickbclifford It seems unlikely that Prettier will change its behavior here. But a simple workaround is to use the CSS pseudo-classes [`:is`](https://developer.mozilla.org/docs/Web/CSS/:is) and [`:where`](https://developer.mozilla.org/docs/Web/CSS/:where) (they only differ in specificity).

Prettier will not break inside these selectors:

```css
:is(h1, h2, h3, h4, h5, h6) {
  margin: 0;
}
```

If your selectors have a common parent or child, I wouldn't even consider this a workaround but the new best practice:

**Old CSS**
```css
section h1,
section h2,
section h3,
section h4,
section h5,
section h6 {
  margin: 0;
}
````

**New CSS**
```css
section :is(h1, h2, h3, h4, h5, h6) {
  margin: 0;
}
```

Browser support is 90% for [`:is`](https://caniuse.com/css-matches-pseudo) and 81% for [`:where`](https://caniuse.com/?search=:where) so unless you still care about IE, you can start using at least `:is` right away. 😎

**Comment by Shirobachi** [comment]
It looks like I've the same problem..
I've
```css
      p, h3, ul {
        margin: 0;
        padding: 0;
      }
```

Prettier make:
```css
      p,
      h3,
      ul {
        margin: 0;
        padding: 0;
      }
```

I want to have:
```
      p, h3, ul {
        margin: 0;
        padding: 0;
      }
```

Is it possible?

**Comment by AlexandroPerez** [comment]
@Jason-2020 I was having the same issue, and was about to have all my `.css` files be ignored by prettier, when I stumbled upon the [ignore comment](https://prettier.io/docs/en/ignore.html#css), when I was looking for a way to do it.

@Shirobachi So we can have the following in css:

```css
/* prettier-ignore */
p, h3, ul {
  margin: 0;
  padding: 0;
}
```

@janosh and the following in javascript:
```javascript
import styled from "styled-components";

import mediaQuery from "../../utils/mediaQuery";

// prettier-ignore
const Participants = styled.section`
  padding: 2rem;
  background: ${props => props.theme.mainWhite};
  display: grid;
  grid-gap: 2rem;
  h1, h2, h3, h4 {
    width: 100%;
    padding: 0.5rem 2rem;
    background: ${props => props.theme.lightGreen};
  }
  ${mediaQuery.phone} {
    h1, h2, h3, h4 {
      font-size: 1rem;
    }
  }
  ${mediaQuery.tablet} {
    grid-gap: 1rem;
    h1, h2, h3, h4 {
      grid-column: 1/-1;
    }
  }
  ${mediaQuery.minTablet} {
    h1, h2, h3, h4 {
      white-space: nowrap;
      grid-row: 2;
    }
  }
`

export default Participants;

```
Is a bit annoying to have to add a comment, and it has to be added before every block of code, or **_abstract syntax tree_** as referred to by prettier docs, that is to be excluded from formatting.

**Comment by bdlowery** [comment]
> @Jason-2020 I was having the same issue, and was about to have all my `.css` files be ignored by prettier, when I stumbled upon the [ignore comment](https://prettier.io/docs/en/ignore.html#css), when I was looking for a way to do it.


THANK YOU. Now my meyer-reset doesn't look like complete shit every-time I add it to a project.

before (disgusting)
<img width="372" alt="Pasted_Image_7_23_21__11_54_PM" src="https://user-images.githubusercontent.com/16507857/126858943-c6c0ec6b-2133-4b6c-97ed-07a5d74f6de2.png">

after (better)
<img width="375" alt="Pasted_Image_7_23_21__11_52_PM" src="https://user-images.githubusercontent.com/16507857/126858925-ffaf8721-0583-4804-9dd8-19fb7cd1fa05.png">

**Comment by arimgibson** [comment]
Just a thought, has anyone tried to PR this idea? I wasn't able to find anything and 10000% understand the want to make sure a PR will be accepted before writing the code for it, but I wonder if that's the main thing holding this up. Clearly there's support for this being a setting, which also means those who are fans of the current use could still have their way. Even keep it as the default.

**Comment by NLEPINAY** [comment]
Oh dear god.

What it's meant to be:

```scss

$gold-gradient: radial-gradient(circle, rgba(222, 208, 160, 1) 0%, rgba(167, 150, 102, 1) 100%);
```

How it turns out:
```scss

$gold-gradient: radial-gradient(circle,
rgba(222,
208,
160,
1) 0%,
rgba(167,
150,
102,
1) 100%);
```

Adding a /* prettier-ignore */ above the line doesn't work for me.

**Comment by ShahriarKh** [comment]
Any updates?

**Comment by uzbeki** [comment]
any updates? Beautify extension already does this.

**Comment by heppokofrontend** [comment]
Similar problems
https://github.com/prettier/prettier/issues/15070

**Comment by lethargistic** [comment]
any updates on this in 2026?
would highly appreciate the option

**Comment by bdlowery** [comment]
> any updates on this in 2026? would highly appreciate the option

yes. for 2026 these are your best options: https://openai.com/codex/ or https://code.claude.com/docs/en/overview#get-started or https://cursor.com/

Then you won't need to worry about these inconsequential problems anymore
